File: c++/november20_onwards/Codechef/novlc2.cpp
#include<bits/stdc++.h>
#define fastio ios_base::sync_with_stdio(false);cin.tie(0);cout.tie(0);
#define endl "\n"
#define ll long long int
#define tc long t;cin>>t;while(t--)
#define pb push_back
#define vi vector<int>
#define vc vector<char>
#define vll vector<ll>
#define umap unordered_map
using namespace std;
int main()
{
	fastio;

  	#ifndef ONLINE_JUDGE
    	freopen("input1.txt","r", stdin);
   		freopen("output1.txt","w", stdout);
	#endif
tc
{
	int n;cin>>n;
	vi b(n+1,0),a(n+1,0);
	int m=3;a[n]=1;
	for(int i=1;i<=n;i++)cin>>b[i];
		for(int i=n-1;i>=1;i--)
	    {
	        if(b[i]==i) a[i]=m++;
	        else a[i]=a[b[i]];
	    }
	for(int i=1;i<=n;i++)
		cout<<a[i]<<" ";
	cout<<endl;
}
}

File: c++/november20_onwards/Codechef/novlc2.py
file1=open("input1.txt","r")
file2=open("output1.txt","w")
t= int(input())
while t:
	t-=1
	n= int(input())
	b=list(map(int, input().split()))
	b.insert(0,0)
	a=[0]*(n+1)
	a[n]=1
	m=3
	for i in range(n-1,0,-1):
		if(b[i]==i):
			a[i]=m
			m+=1
		else:
			a[i]=a[b[i]]
	for i in range(1,n+1):
		print(a[i],end=" ")
	print()
